benches/criterion.rs

```diff
@@ -0,0 +1,271 @@
+        b.iter(move || {
+            let options = naga::front::spv::Options::default();
+            for input in inputs.iter() {
+                let spv =
+                    unsafe { slice::from_raw_parts(input.as_ptr() as *const u32, input.len() / 4) };
+                let parser = naga::front::spv::Parser::new(spv.iter().cloned(), &options);
+                parser.parse().unwrap();
+            }
+        });
+    });
+    #[cfg(feature = "glsl-in")]
+    group.bench_function("glsl", |b| {
+        let vert = gather_inputs("tests/in/glsl", "vert");
+        b.iter(move || parse_glsl(naga::ShaderStage::Vertex, &vert));
+        let frag = gather_inputs("tests/in/glsl", "frag");
+        b.iter(move || parse_glsl(naga::ShaderStage::Vertex, &frag));
+        //TODO: hangs for some reason!
+        //let comp = gather_inputs("tests/in/glsl", "comp");
+        //b.iter(move || parse_glsl(naga::ShaderStage::Compute, &comp));
+    });
+}
+
+#[cfg(feature = "wgsl-in")]
+fn gather_modules() -> Vec<naga::Module> {
+    let inputs = gather_inputs("tests/in", "wgsl");
+    let mut parser = naga::front::wgsl::Parser::new();
+    inputs
+        .iter()
+        .map(|input| {
+            let string = std::str::from_utf8(input).unwrap();
+            parser.parse(string).unwrap()
+        })
+        .collect()
+}
+#[cfg(not(feature = "wgsl-in"))]
+fn gather_modules() -> Vec<naga::Module> {
+    Vec::new()
+}
+
+fn validation(c: &mut Criterion) {
+    let inputs = gather_modules();
+    let mut group = c.benchmark_group("valid");
+    #[cfg(feature = "validate")]
+    group.bench_function("safe", |b| {
+        let mut validator = naga::valid::Validator::new(
+            naga::valid::ValidationFlags::all(),
+            naga::valid::Capabilities::all(),
+        );
+        b.iter(|| {
+            for input in inputs.iter() {
+                validator.validate(input).unwrap();
+            }
+        });
+    });
+    #[cfg(feature = "validate")]
+    group.bench_function("unsafe", |b| {
+        let mut validator = naga::valid::Validator::new(
+            naga::valid::ValidationFlags::empty(),
+            naga::valid::Capabilities::all(),
+        );
+        b.iter(|| {
+            for input in inputs.iter() {
+                validator.validate(input).unwrap();
+            }
+        });
+    });
+}
+
+fn backends(c: &mut Criterion) {
+    #[cfg(feature = "validate")]
+    let inputs = {
+        let mut validator = naga::valid::Validator::new(
+            naga::valid::ValidationFlags::empty(),
+            naga::valid::Capabilities::empty(),
+        );
+        let input_modules = gather_modules();
+        input_modules
+            .into_iter()
+            .flat_map(|module| validator.validate(&module).ok().map(|info| (module, info)))
+            .collect::<Vec<_>>()
+    };
+    #[cfg(not(feature = "validate"))]
+    let inputs = Vec::<(naga::Module, naga::valid::ModuleInfo)>::new();
```

Comment:
Again, we shouldn't run benchmarks on empty input lists because a feature was turned off.

Reply:
similar reasoning to the last one - this is the minimal change to compile